File: server_python/src/services/post_service.py
# -*- coding: utf-8 -*-
from typing import Union, List

import src.repositories.post_repository as post_repo
from src.models.inputs.add_post_args import AddPostArgs
from src.models.payloads.bool_payload import BoolPayload
from src.models.payloads.error_item import ErrorItem
from src.models.payloads.post_payload import PostPayload
from src.models.payloads.posts_payload import PostsPayload
from src.utils.validations import validate_numeric_args


def _validate_add_post_args(data: AddPostArgs) -> Union[List[ErrorItem], None]:
    if data is None:
        return [ErrorItem("Input data cannot be null.")]

    if data.author_id < 0:
        return [ErrorItem("Author Id must be provided and must be valid.")]

    if data.title is None or data.title.strip() == "":
        return [ErrorItem("All posts must have a title.")]

    if data.content is None or data.content.strip() == "":
        return [ErrorItem("All posts must have a content.")]

    return None


def post_get_paginated(skip: int, take: int) -> PostsPayload:
    errors = validate_numeric_args({"skip": skip, "take": take}, 0)
    if errors is not None:
        return PostsPayload(errors=errors)

    try:
        return PostsPayload(posts=post_repo.post_get_paginated(skip=skip, take=take))
    except Exception as e:
        return PostsPayload(errors=[ErrorItem(f"Failed to get paginated posts: {e}")])


def post_get_paginated_by_author(author_id: int, skip: int, take: int) -> PostsPayload:
    errors = validate_numeric_args({"author_id": author_id, "skip": skip, "take": take}, 0)
    if errors is not None:
        return PostsPayload(errors=errors)

    try:
        return PostsPayload(posts=post_repo.post_get_paginated_by_author(author_id=author_id, skip=skip, take=take))
    except Exception as e:
        return PostsPayload(errors=[ErrorItem(f"Failed to get paginated posts by author: {e}")])


def post_insert(data: AddPostArgs) -> PostPayload:
    errors = _validate_add_post_args(data)
    if errors is not None:
        return PostPayload(errors=errors)

    try:
        return PostPayload(post=post_repo.post_insert(data))
    except Exception as e:
        return PostPayload(errors=[ErrorItem(f"Failed to insert post: {e}")])


def post_delete(post_id: int) -> BoolPayload:
    if post_id < 0:
        return BoolPayload(errors=[ErrorItem("A Post Id must be provided and must be a valid number.")])

    try:
        return BoolPayload(result=post_repo.post_delete(post_id))
    except Exception as e:
        return BoolPayload(errors=[ErrorItem(f"Failed to insert post: {e}")])


def post_get_by_id(post_id: int) -> PostPayload:
    if post_id < 0:
        return PostPayload(errors=[ErrorItem("A Post Id must be provided and must be a valid number.")])

    try:
        return PostPayload(post=post_repo.post_get(post_id))
    except Exception as e:
        return PostPayload(errors=[ErrorItem(f"Failed to get post by id: {e}")])

File: server_nodejs/src/services/post-service.ts
import * as postRepo from "../repositories/post-repository";
import {AddPostArgs, PostPayload, PostService} from "../interfaces/post-interfaces";
import {BooleanPayload} from "../interfaces/boolean-interfaces";

const validateAddPostArgs = (data: AddPostArgs):PostPayload | null => {
    if (!data)
        return {errors: [{message: "Input data cannot be null."}],}

    if (!data.authorId || data.authorId < 0)
        return {errors: [{message: "Author Id must be provided and must be valid."}],}

    if (!data.title)
        return {errors: [{message: "All posts must have a title."}],}

    if (!data.content)
        return {errors: [{message: "All posts must have content."}],}

    return null;
}

const postInsert = (data: AddPostArgs): PostPayload => {
    const error = validateAddPostArgs(data);
    if (error) return error;

    try {
        return {
            errors: [],
            post: postRepo.postInsert(data)
        };
    } catch (e) {
        return {
            errors: [{message: "Failed to add Post." + e}],
        }
    }
};

const postDelete = (postId: number): BooleanPayload => {
    if (!postId || Number.isNaN(postId) || postId < 0)
        return {
            errors: [{message: "A Post Id must be provided and must be a valid number."}],
            result: false
        }

    try {
        return {
            errors: [],
            result: postRepo.postDelete(postId)
        };
    } catch (e) {
        return {
            errors: [{message: "Failed to delete Post." + e}],
            result: false
        }
    }
}

const postGetPaginated = (skip: number = 0, take: number = 100): PostPayload => {
    if (skip < 0 || take < 0)
        return {
            errors: [{message: "Skip and Take parameters must be greater than zero."}]
        }

    try {
        return {
            errors: [],
            posts: postRepo.postGetPaginated(skip, take)
        }
    } catch (e) {
        return {
            errors: [{message: "Failed to get paginated Posts." + e}],
        }
    }
}

const postGetPaginatedByAuthor = (authorId: number, skip: number = 0, take: number = 100): PostPayload => {
    if (!authorId || authorId < 0)
        return {
            errors: [{message: "Author Id must not be null and must be greater than zero."}]
        }

    if (skip < 0 || take < 0)
        return {
            errors: [{message: "Skip and Take parameters must be greater than zero."}]
        }

    try {
        return {
            errors: [],
            posts: postRepo.postGetPaginatedByAuthor(authorId, skip, take)
        }
    } catch (e) {
        return {
            errors: [{message: "Failed to get paginated Posts by author." + e}],
        }
    }
}

const postGetByPostId = (postId:number): PostPayload => {
    if (!postId || postId < 0)
        return {
            errors: [{message: "Post Id must not be null and must be greater than zero."}]
        }

    try {
        return {
            errors: [],
            post: postRepo.postGetByPostId(postId)
        }
    } catch (e) {
        return {
            errors: [{message: "Failed to get post by post id." + e}],
        }
    }
}

export const postService: PostService = {
    postInsert,
    postDelete,
    postGetPaginated,
    postGetPaginatedByAuthor,
    postGetByPostId
}